File: go/misarblog.go
package misarblog

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

const EMBED_BASE = "https://misar.blog"

type TokenResult struct {
	Token     string `json:"token"`
	ExpiresAt int64  `json:"expiresAt"`
}

func EmbedURL(username, slug, theme string) string {
	var path string
	if slug != "" {
		path = fmt.Sprintf("%s/%s/%s/embed", EMBED_BASE, username, slug)
	} else {
		path = fmt.Sprintf("%s/%s/embed", EMBED_BASE, username)
	}
	if theme != "" && theme != "auto" {
		path = fmt.Sprintf("%s?theme=%s", path, theme)
	}
	return path
}

func RefreshToken(token, baseURL string) (*TokenResult, error) {
	payload, err := json.Marshal(map[string]string{"token": token})
	if err != nil {
		return nil, err
	}
	resp, err := http.Post(baseURL+"/api/auth/refresh", "application/json", bytes.NewBuffer(payload))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("refresh failed: %d", resp.StatusCode)
	}
	var result TokenResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: cpp/src/misarblog.cpp
#include "misarblog.hpp"
#include <curl/curl.h>
#include <nlohmann/json.hpp>
#include <stdexcept>
#include <sstream>

using json = nlohmann::json;

namespace misarblog {

static size_t writeCallback(void *contents, size_t size, size_t nmemb, std::string *s) {
    s->append(static_cast<char *>(contents), size * nmemb);
    return size * nmemb;
}

std::string embedUrl(const std::string &username, const std::string &slug, const std::string &theme) {
    std::string url = EMBED_BASE + "/" + username;
    if (!slug.empty()) url += "/" + slug;
    url += "/embed";
    if (!theme.empty() && theme != "auto") url += "?theme=" + theme;
    return url;
}

TokenResult refreshToken(const std::string &token, const std::string &baseUrl) {
    std::string url = baseUrl + "/api/auth/refresh";
    std::string body = json{{"token", token}}.dump();
    std::string response;

    CURL *curl = curl_easy_init();
    if (!curl) throw std::runtime_error("curl_easy_init failed");

    struct curl_slist *headers = curl_slist_append(nullptr, "Content-Type: application/json");
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

    CURLcode res = curl_easy_perform(curl);
    long status = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK) throw std::runtime_error(curl_easy_strerror(res));
    if (status != 200) throw std::runtime_error("HTTP error: " + std::to_string(status));

    auto j = json::parse(response);
    return TokenResult{j.at("token").get<std::string>(), j.at("expiresAt").get<long long>()};
}

} // namespace misarblog
